hw2/curry.py

```diff
@@ -0,0 +1,65 @@
+import inspect
+from functools import wraps, update_wrapper
+
+from typing import Callable, Any, TypeVar
+
+
+def positional_arity(func: Callable[..., Any]) -> int:
+    """
+    Returns the number of positional arguments for func (excluding varargs).
+    :param func: any function
+    :return: number of positional arguments of func
+    """
+    args, *rest = inspect.getfullargspec(func)
+    return len(args)
+
+
+def has_varargs(func: Callable[..., Any]) -> bool:
+    """
+    :param func: any function
+    :return: True if func accepts varargs else False
+    """
+    _, vararg, *rest = inspect.getfullargspec(func)
+    return vararg is not None
+
+
+R = TypeVar("R")
+
+
+def curry_internal(func: Callable[..., R], arity: int, passed_args: list = None) -> Callable:
+    """
+    For internal use by curry_explicit only!
+    """
+    if passed_args is None:
+        passed_args = []
+
+    actual_arity = positional_arity(func)
+    if (not has_varargs(func) and arity != actual_arity) or arity < actual_arity:
+        raise TypeError(
+            f"Received function {func} with {actual_arity} positional arguments but arity was less: {arity}"
+        )
+
+    if arity == 0:
+        return update_wrapper(lambda: func(), func)
+
+    @wraps(func)
+    def curried_function(x):
+        if len(passed_args) + 1 < arity:
+            return curry_internal(func, arity, [*passed_args, x])
+        else:
+            return func(*passed_args, x)
+
+    return curried_function
+
+
+def curry_explicit(func: Callable[..., R], arity: int) -> Callable:
```

Comment:
Please, try to fix your functions for this case:
```
f4 = curry_explicit(print, 4)
g4 = uncurry_explicit(f4, 4)
f2 = f4(10)(20)
g2 = uncurry_explicit(f2, 2)
print('---')
f4(5)(15)(25)(35)
f2(30)(40)
print('---')
g4(5, 15, 25, 35)
g2(30, 40)
```
The expected output in this case:
```
---
5 15 25 35
10 20 30 40
---
5 15 25 35
10 20 30 40
```
But in your case, you will have an error

Reply:
For some reason python raises an exception when I try to inspect the signature of the `print` function